### ISSUE TITLE
Rename SSH3 => SSHH3 ? 

### ISSUE THREAD

**Issue by telamon** [francoismichel/ssh3#79]
This project is not an iteration of SSH as it does not extend the SSH protocol itself and is not compatible with existing clients nor tooling and runs on a completely different port and server architecture.
Which leads me to assume that the number `3` was acquired from the use of http/3 proposals.

As such the long form description is **"SSHv2 over HTTP/3"** so my question is:
Wouldn't it be more correct to shorten it to `sshh3` or `sshh` ?

-- You've created something cool but completely different :-)

**Comment by nileshpatra** [comment]
/cc: @francoismichel it'd be great to have your opinion on this.

**Comment by stappersg** [comment]
Short:  Please do `s/ssh3/sshh3/g`,  global substitute `"ssh3"` with `"sshh3"`.


Several years ago there was _use SSH2, because SSH1 has security flaws_. Today I heard about `ssh3` and was thinking _"O chips, need to update all SSH2 stuff"_. Then I learnt this actual ssh**h3**.  `:-/`

The _it is a poor name_ is also expressed in https://lists.debian.org/debian-devel/2023/12/msg00213.html and it's follow-up messages.

And in https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1059618#26 is stated:
> I share these concerns, so I'll delay the upload for now.  I'm hoping
upstream will rename the project to something less confusing.

**Comment by mikelolasagasti** [comment]
Another option could be `soh` for "SSH over HTTP" or "Shell over HTTP".

Waiting on the final decision to package this for Fedora: https://bugzilla.redhat.com/show_bug.cgi?id=2256027

**Comment by stappersg** [comment]
> Please do `s/ssh3/sshh3/g`, global substitute `"ssh3"` with `"sshh3"`.

It became `h3sh`, inspired by https://lists.debian.org/debian-devel/2023/12/msg00225.html
> something as simple as naming it h3sh would have avoided the brand confusion 

There are merge request #85, #86 and #87. That last one:
```text
stappers@paddy:~/xkcd386/h3sh
$ git show
commit 22b6ac4b4be0f8203e6103ef8c4ebe038239f5ba
Author: Geert Stappers <stappers@stappers.it>
Date:   Sun Dec 31 13:38:15 2023 +0100

    Explaining the name.
    
    Because we can't afford ourselfs the effort to explain
    all the SSH users that this is not SSH3.
    
    Signed-off-by: Geert Stappers <stappers@stappers.it>

diff --git a/README.md b/README.md
index bb767ae..a81f99b 100644
--- a/README.md
+++ b/README.md
@@ -19,7 +19,7 @@ Among others, H3SH allows the following improvements:
 > [!TIP]
 > Quickly want to get started ? Checkout how to [install H3SH](#installing-h3sh). You will learn to
 
-*H3SH* stands for the concatenation of *SSH* and *H3*.
+*H3SH* stands for *HTTP/3* and *shell*.
 
 ## ⚡ H3SH is faster
 Faster for session establishment, not throughput ! H3SH offers a significantly faster session
stappers@paddy:~/xkcd386/h3sh
$
```

And regarding the _"appears to be intentional"_ from
> Even something as simple as naming it h3sh would have avoided the brand confusion while communicating the purpose of the package. This does not appear to be a case of "unknowing infringement." It appears to be intentional.

Younger versions of me also **did intentional** _"this will be funny"_ and got back _"not funny"_ that made me better.

**Comment by jas4711** [comment]
IMHO typing `h3sh` at the command line is not that pleasant.  I suggest not using a digit in the name.  Let's use a name that is unique and where typing and tab-completion is simple and package naming doesn't conflict or confuse with existing projects.

**Comment by stappersg** [comment]
On Sun, Dec 31, 2023 at 05:46:24AM -0800, Simon Josefsson wrote:
> IMHO typing `h3sh` at the command line is not that pleasant.  I suggest
> not using a digit in the name.  Let's use a name that is unique and
> where typing and tab-completion is simple and package naming doesn't
> conflict or confuse with existing projects.

s/3/e/   gives   hesh.  Yes, much more pleasant to type as h3sh.

https://duckduckgo.com/?q=hesh  didn't show conflicts with software
projects, nor with SSH.


-*H3SH* stands for the concatenation of *SSH* and *H3*.
+*HESH* stands for the 3 in *HTTP/3* mirrorerd into E and *shell*.



Happy New Year
Geert Stappers
-- 
Silence is hard to parse

**Comment by francoismichel** [comment]
Sorry for the delay, I am currently out of office.

I have no problem with rethinking the name, and I am also okay with distros renaming the binaries 
however they want right now. I am happy with people discussing this here and I take note of the suggestions. We however may want to wait a little bit before applying renaming on the whole project. I think the question about naming will become clearer once we get to discuss the project at IETF in March. We discussed 
a few alternative name ideas. shh is one of the candidates that I like, the only thing I don’t like in it is that it looks like an ssh typo. shs can be cool as well.
I am a bit less convinced with names with the « 3 » in it as one of the goals could also be to make it support HTTP/2 (again, this might become clearer after a few IETF discussions).

I think waiting until March-April would help us coming up with a more solid name, and that would also give time to other people for discussing names here as well. In the mean time, distros naming it how they want if they don’t like ssh3 is okay to me, the packages would probably be proposed for unstable distribution releases anyway.

Meanwhile, Happy New Year and thank you for the thoughtful comments and the momentum you give to the project ! I am really looking forward to the release 0.1.5 that includes a lot of cool improvements that will make it great for day-to-day use :-)

I let the issue open right now as I would like to keep getting more feedback in the meantime.

**Comment by iustin** [comment]
My 2 bits here as a random small open source developer. March is 3 months away. Keeping ssh3 for another 3 months especially as "3" should not be in the name (per your comment) doesn't make sense.

Just rename it to a random, even long name for now - ssh-over-http3, and in March once the future goals are clearer, find a good, proper name.

**Comment by jas4711** [comment]
Maybe if someone can create a pull request with ONE proposed naming, and if we distributors can all agree to use that, pending a final decision?  I'm preparing packages for Debian.  If some other packagers can agree on a name, we can get consistency even if upstream wants to delay the decision (which is entirely reasonable).

I've done some typing tests, and my prefer is `shs` because it is fast to type on a keyboard.  Repeating the same character on a keyboard as in `ssh` or `shh` is slower than alternating characters like `shs`.  The acronym would be Secure Hypertext Shell which actually makes a lot of sense.

Thoughts?  For packaging purposes, I think it is sufficient to rename the binaries /usr/bin/shs and /usr/bin/shsd, as I'm thinking the ssh3-server binary would eventually be extended to implement daemon-like functionality.  I think shsd should not be in /usr/sbin as users may want to run it too, as a non-root setup. Project documentation and golang name stays the same.  Thoughts?

**Comment by francoismichel** [comment]
> Maybe if someone can create a pull request with ONE proposed naming, and if we distributors can all agree to use that, pending a final decision? I'm preparing packages for Debian. If some other packagers can agree on a name, we can get consistency even if upstream wants to delay the decision (which is entirely reasonable).

I like this idea, enabling us to have common package names without renaming in a hurry. Time will also tell us if the common package names is a good candidate for being the potential long-term name. I like `soh` and `shs`, I am just a little scared of `shs` being close to an `ssh` typo. It might be worth having the opinion if several people involved in distros packaging.

**Comment by jas4711** [comment]
> I like `soh` and `shs`, I am just a little scared of `shs` being close to an `ssh` typo. It might be worth having the opinion of several people involved in distros packaging.

Yeah, the acronym for `shs` is less good as @CameronNemo noted.  I did some typing tests for `soh` and for me it involves using either three fingers (which cognitively takes more resources than two fingers) or movement of one finger (which also takes resources).   Overall `shs` feels faster to type though.  Yes, I am doing premature optimizations =)

**Comment by jas4711** [comment]
> Iterating on this, what do we explicitly need in such pull request for distro packages ? I am not sure we need to rename every file, every object and package, we could stick to changing the binaries names, logging and readme for now, that would make the PR a lot lighter and probably a lot easier to maintain up-to-date with upstream.
> 
> Thoughts ?

- [ ] Rename binaries /usr/bin/ssh3 and /usr/bin/ssh3-server
- [ ] Rename ~/.ssh3/ naming?  for consistency with the binaries
- [ ] ?

I think settling on naming is going to be the main challenge to get this packaged, at least for Debian.  Doing an upload, even to experimental, only to later have to rename it requires some manual steps, so I'm reluctant to do that just now.

**Comment by iustin** [comment]
I don't think the optimisation here should be fast typing. `shs` or `soh` is almost the same. Yes, there's a difference, but IMO it's way more likely that one will mistype `ssh` as `shs` or viceversa, than be bothered by the slow `soh` typing.

I would vote for moving ahead with `soh`.

**Comment by jas4711** [comment]
> I don't think the optimisation here should be fast typing. `shs` or `soh` is almost the same. Yes, there's a difference, but IMO it's way more likely that one will mistype `ssh` as `shs` or viceversa, than be bothered by the slow `soh` typing.
> 
> I would vote for moving ahead with `soh`.

Works for me if we can agree on it -- any other packagers here?  Which distributions are interested?

**Comment by mikelolasagasti** [comment]
Fedora and potentially EPEL packager here, fine with `soh` (I proposed it :sweat_smile:).

**Comment by francoismichel** [comment]
Great. I can make an `soh` release based on version 0.1.5-rc3 (or rc4). I think the new ProxyJump support (#44) makes it more usable so let's include it in the release.

**Comment by mpiraux** [comment]
I'm very fine with `soh` for all the reasons discussed here :)

**Comment by stappersg** [comment]
There is now #96 which has

> Two main reasons:
>   Dropped the 3 to be ready for HTTP/2
>   Avoiding confusion with OpenSSH. ( We don't want to panic SSH users to update their SSH. And we don't want to keep explaining this project is not next OpenSSH release.)

in the commit message.

Regarding
> Iterating on this, what do we explicitly need in such pull request for distro packages ? I am not sure we need to rename every file, every object and package, we could stick to changing the binaries names, logging and readme for now, that would make the PR a lot lighter and probably a lot easier to maintain up-to-date with upstream.
> 
> Thoughts ?

I think the name "ssh3" did serve it's purpose, it did bring this project attention.  Now the project deserves a better name.  The sooner, the better.

**Comment by telamon** [comment]
Reverse it?
`hss`

**Comment by francoismichel** [comment]
Thanks for the PR, I'll have a look at it at the beginning of next week.
As I said, I am fine if the binaries get named something else than `ssh3`. For the *protocol* in itself, we'll see after a few IETF discussions if it becomes a version of SSH or something else, but it will probably be presented as an SSH version: it really implements the SSH connection protocol.

> Reverse it?
> hss

We kind of agreed on a package name but I am okay if people want to continue to propose, I just think we'll stick to `soh` for packaging right now. It would be great to have a place easing propositions and votes.

**Comment by rbalint** [comment]
> Reverse it? `hss`

Think this is quite good, HTTP Secure Shell, or HTTPS Shell.

**Comment by ledlamp** [comment]
hsh

**Comment by stappersg** [comment]
![commits](https://github.com/francoismichel/ssh3/assets/35434094/8cf93147-bb98-4b8a-a33e-1e03a9ae3b74)

**Comment by toomim** [comment]
I like `soh`, and also "QUIC Shell", or `qsh`.

This extends the `rsh` → `ssh` → `qsh` etymology.  
From *Remote* Shell, to *Secure* Shell, to *QUIC* Shell.

**Comment by almereyda** [comment]
While `soh` leans in to `doh`/DoH, DNS-over-HTTPS, it might be less known to people than their shell.

The string from `bash` to `ash` and `zsh` via `mosh` and `ssh` leading to `qsh` on the other side shows applicapility in familiar environments. Strong favourite. And a QUIC Shell is what it is, is it not?

**Comment by Splizard** [comment]
How about `shhh`?

The 3 `h`'s referencing HTTP *3* and the fact that it's invisible to port scanning.

**Comment by toomim** [comment]
^ That's hilarious. Which makes it memorable, which is very important for a good name.
